### PR TITLE
find_new_[ug]id: Skip over IDs that could overflow to -1

### PR DESCRIPTION
**Pull request by nertpinx**
Due to compatibility reasons neither UID nor GID should overflow to -1, be it on systems with 16-bit or 32-bit uid_t/gid_t since -1 is special.  It has different meaning for example as a chown(2) parameter, could become 65535 in programs compiled when uid_t was 16bit and possibly more issues.

### CODE REVIEW

File: libmisc/find_new_gid.c
Comment (ikerexxe) [inline libmisc/find_new_gid.c:116]: Indentation issue

File: libmisc/find_new_gid.c
Comment (ikerexxe) [inline libmisc/find_new_gid.c:195]: You should update this comment as you added a new case

File: libmisc/find_new_gid.c
Comment (ikerexxe) [inline libmisc/find_new_gid.c:117]: Where is `u_int16_t` defined? Maybe it's a typo for `uint16_t`?

**Review by ikerexxe — CHANGES_REQUESTED** [review]
Added some comments that need to be addressed.

File: libmisc/find_new_gid.c
Comment (alejandro-colomar) [reply libmisc/find_new_gid.c:117]: BTW, you could get those values as `UINT16_MAX` and `UINT32_MAX` and avoid any casts.

Comment (ikerexxe) [reply libmisc/find_new_gid.c:117]: That value (65535) is also reserved in 32-bit systems, and several applications have problems handling it. As an example there's [systemd](https://systemd.io/UIDS-GIDS/), check the `Special Linux UIDs` section.

Comment (alejandro-colomar) [reply libmisc/find_new_gid.c:117]: Hmmm, makes sense.  Then I suggest using `UINT16_MAX` and `UINT32_MAX`.

File: libmisc/find_new_gid.c
Comment (nertpinx) [reply libmisc/find_new_gid.c:116]: Sorry, I did not set up everything properly looks like. Is the preferred default the same as with linux (indentation with tabs of width 8)?

File: libmisc/find_new_gid.c
Comment (nertpinx) [reply libmisc/find_new_gid.c:195]: I rather went ahead and adjusted all of them. It looked like we might not need to differentiate between `EEXIST` and `EINVAL`, but I went for the safer route.

File: libmisc/find_new_gid.c
Comment (nertpinx) [reply libmisc/find_new_gid.c:117]: To be honest I have no idea, probably just a completion typo or something. Looks like it is both in glibc headers and in linux, but It was meant to be `uintX_t`.

Comment (nertpinx) [reply libmisc/find_new_gid.c:117]: Sure, I can switch to that, it's just that the -1 cast is more intuitive that I went for that instead.

File: libmisc/find_new_gid.c
Comment (ikerexxe) [reply libmisc/find_new_gid.c]: I don't think we have any preference, we just try to keep the same indentation throughout the file.

**Review by ikerexxe — APPROVED** [review]
LGTM, but as I provided the general idea for the solution I'd like somebody else to review it before merging. @alejandro-colomar wdyt?

**Comment by alejandro-colomar** [comment]
LGTM, but I'd add something more to the commit message.  Having compatibility to 16-bit programs seems crazy these days, IMO.  I don't think distros ship those anymore.  The real issue seems to be that some other (64-bit) programs have had compatibility to 16-bit programs (and keep that compatibility even today), and so we need to be compatible with those programs that are compatible to 16-bit programs.  And a link to systemd's docs would also help.

**Comment by alejandro-colomar** [comment]
> > LGTM, but I'd add something more to the commit message. Having compatibility to 16-bit programs seems crazy these days, IMO. I don't think distros ship those anymore. The real issue seems to be that some other (64-bit) programs have had compatibility to 16-bit programs (and keep that compatibility even today), and so we need to be compatible with those programs that are compatible to 16-bit programs. And a link to systemd's docs would also help.
> 
> Some IOT devices however might?

But the problem is not 16-bit id_t, because that should already be caught by the normal range check (I guess that would result in `ERANGE`).  The problem is 16-bit applications in 32-bit systems, or 32-bit applications in 64-bit systems, if I understood this right.

I don't expect IOT devices to have these problems (although they might).  I expect those problems to only exist in very old archs such as x86, which has many traces of legacy stuff.  Whatever it is, I'd like it to be documented in the commit message, so that 10 years from now whoever picks the code knows why those weird tests are there. :)

**Comment by hallyn** [comment]
> > > LGTM, but I'd add something more to the commit message. Having compatibility to 16-bit programs seems crazy these days, IMO. I don't think distros ship those anymore. The real issue seems to be that some other (64-bit) programs have had compatibility to 16-bit programs (and keep that compatibility even today), and so we need to be compatible with those programs that are compatible to 16-bit programs. And a link to systemd's docs would also help.
> > 
> > 
> > Some IOT devices however might?
> 
> But the problem is not 16-bit id_t, because that should already be caught by the normal range check (I guess that would result in `ERANGE`). The problem is 16-bit id_t applications in 32-bit id_t systems, or 32-bit id_t applications in 64-bit id_t systems, if I understood this right.

I think you're right.

**Comment by nertpinx** [comment]
> > > > LGTM, but I'd add something more to the commit message. Having compatibility to 16-bit programs seems crazy these days, IMO. I don't think distros ship those anymore. The real issue seems to be that some other (64-bit) programs have had compatibility to 16-bit programs (and keep that compatibility even today), and so we need to be compatible with those programs that are compatible to 16-bit programs. And a link to systemd's docs would also help.
> > > 
> > > 
> > > Some IOT devices however might?
> > 
> > 
> > But the problem is not 16-bit id_t, because that should already be caught by the normal range check (I guess that would result in `ERANGE`). The problem is 16-bit id_t applications in 32-bit id_t systems, or 32-bit id_t applications in 64-bit id_t systems, if I understood this right.
> 
> I think you're right.

Yes, that's the issue, at least how I understood it. I can add a link to https://systemd.io/UIDS-GIDS/ in the commit message to make it clear I did not come up with it out of nowhere ;)

**Comment by alejandro-colomar** [comment]
I guess that's enough, but let me suggest some alternative text:

```
Some programs don't support `(uint16_t) -1` or `(uint32_t) -1` as user
or group IDs.  This is because `-1` is used as an error code, and in the
past, `gid_t` and `uid_t` have changed width.  For legacy reasons, those
values have been kept reserved in programs today (for example systemd
does this; see the documentation in the link below).

This should not be confused with catching overflow in the ID values,
since that is already caught by our ERANGE checks.  This is about not
using reserved values that have been reserved for legacy reasons.

Link: <https://systemd.io/UIDS-GIDS/>
Cc: ...
Reviewed-by: ...
Reviewed-by: Alejandro Colomar <alx@kernel.org>
Signed-off-by: ...
```

**Comment by nertpinx** [comment]
Thanks for the suggestion, I went with your version, only I took the liberty of adding one more piece of information just for completeness sake in the following sentence:

This is because `-1` is used as an error code or as an unspecified ID, e.g. in chown(2) parameters, and in the past, `gid_t` and `uid_t` have changed width.

**Comment by ikerexxe** [comment]
@alejandro-colomar any additional feedback before I merge?

**Comment by alejandro-colomar** [comment]
LGTM :-)

--
Alejandro Colomar
Linux man-pages comaintainer; https://www.kernel.org/doc/man-pages/
http://www.alejandro-colomar.es/

**Comment by ikerexxe** [comment]
Thanks for the patch!

**Pull request merged**
